### ISSUE TITLE
from_zelig for extracting original fitted model objects

### ISSUE THREAD

**Issue by christophergandrud** [IQSS/Zelig#189]
Building on conversations with @izahn and @vincentarelbundock I've implemented a `from_zelig` method (with a wrapper) that will extract original fitted model objects from `zelig` calls: b761971e6a739d12d97314f63dc902b0d4f44bc4

This can be useful for passing output to non-Zelig post-estimation functions and packages such as texreg and stargazer for creating well-formatted presentation document tables.

It is part of the workflow described in https://github.com/IQSS/Zelig/issues/176

![](https://cloud.githubusercontent.com/assets/1285805/22466918/7db22784-e790-11e6-90f7-7b15c743dfb8.jpg)

**Comment by christophergandrud** [comment]
Minimal example:

```r
z <- zelig(Fertility ~ Education, data = swiss, model = 'ls')
z_out_original <- from_zelig(z)

predict(z_out_original)

##       1        2        3        4        5        6        7        8        9 
## 69.26186 71.84891 75.29831 73.57361 66.67480 73.57361 73.57361 72.71126 73.57361 
##      10       11       12       13       14       15       16       17       18 
## 68.39950 74.43596 69.26186 73.57361 69.26186 75.29831 77.88536 72.71126 55.46425 
##      19       20       21       22       23       24       25       26       27 
## 62.36305 71.84891 70.98656 77.02301 69.26186 74.43596 78.74771 72.71126 77.02301 
##      28       29       30       31       32       33       34       35       36 
## 70.98656 63.22540 72.71126 77.88536 74.43596 77.88536 74.43596 77.02301 71.84891 
##      37       38       39       40       41       42       43       44       45 
## 77.02301 68.39950 69.26186 70.12421 68.39950 52.01485 73.57361 73.57361 33.90549 
##      46       47 
## 54.60190 54.60190
```

**Comment by tercer** [comment]
Since there are presently different uses of the `by` architecture that each result in lists of results, I put some additional messages into this function here:  https://github.com/IQSS/Zelig/commit/ec4162f17f9eb8ef4d323374db4a2d10e8e86bce

One small comment is that since we have a variety of `get*` getters (`getqi`, `getcoef`, `getvcov`, `getpredict`, `getfitted`) would it make sense to call this `getresult` or `getmodel`?

**Comment by izahn** [comment]
I think `getmodel` makes sense for consistency. (I wish we had `get_qi`, `get_coef` etc. instead, but that's another issue.).

The current `from_zelig` makes sense for symmetry with the (proposed but not yet implemented) `to_zelig`, but looking at Christopher's diagram I guess `to_zelig` will be an internal function right? The user will just call `setx`? So consistency/symmetry is less import there.

**Comment by christophergandrud** [comment]
Thanks for those additions @tercer! I had a strong feeling that I wasn't capturing all of the use cases.

The dreaded naming issue...

I've been thinking about this and maybe to avoid confusion, we should think of creating a new series of functions that are considered in a different family from `get*`:

- `from_zelig_model` (currently `from_zelig`  in the diagram)

- `to_zelig_model` (currently `to_zelig`, this could be called internally from `setx` for convenience, but could also be called before it)

- `zelig_qi_to_df` (currently `getqi_df`)

The `get*` functions are largely internal functions (I didn't know about them until digging through the source code). They are generally used to get one part of a Zelig object and pass it to some Zelig function. They also have a more compact naming convention, which seems more suitable to internal code.

The `from`/`to` functions have a different purpose. They are solely aimed at allowing Zelig to interact with the non-Zelig universe. They are more user-facing, so have a more descriptive and less compact naming structure. They use the use `_` to delimit words, like the tidyverse, thus signalling there compatibility with that. (Function names with out word delimitation is harder to read, especially for lower skilled English speakers).

Thoughts?

**Comment by tercer** [comment]
Although we certainly shifted to using them internally, once they started
to become available, the primary purpose of the getters was for users who
were asking how to extract some piece out of the zelig object.  The zelig
object, especially because of the layers added by dplyer and the general
lack of utilities for reference classes, is difficult to navigate.

I can't imagine there are so many people using the getters yet, that we
couldn't shift getthis() to get_this() and then slowly deprecate the first
with a version or two of warning messages.

The from/to from my perspective is fine either way.

**Comment by cchoirat** [comment]
I agree, we should progressively switch to get_this().

**Comment by christophergandrud** [comment]
Sounds good. I'll take a stab at this.

**Comment by christophergandrud** [comment]
`from_zelig_model` with tests for stargazer and texreg requirements added in 93328a81acaeb7fbc237fff89316347a75cecdc5

I'm planning on implementing `to_zelig_model` and `zelig_qi_to_df` in a post 5.0-15 release.

**Comment by christophergandrud** [comment]
Created an experimental implementation of the last part of this workflow: `to_zelig`, which is run inside of `setx` if a non-zelig object is passed to it: https://github.com/IQSS/Zelig/tree/setxdirect

This will now work:

```r
library(zeligverse)
library(dplyr)

m1 <- lm(Fertility ~ Education, 
         weights = sample(1:10, replace = TRUE, size = nrow(swiss)),  
         data = swiss)

setx(m1) %>% sim() %>% plot()
```

The key code is at: https://github.com/IQSS/Zelig/blob/setxdirect/R/interface.R#L1-L33

Note that currently it only works for `lm` and doesn't work with updated formulas, dot formulas, etc.

**Comment by christophergandrud** [comment]
`glm` models for logit, gamma, poisson, probit, (and glm ls): https://github.com/IQSS/Zelig/commit/28662843aa935a61c5b441ffa32edac80df58061

**Comment by christophergandrud** [comment]
Current possible zeligverse workflows as of e1b74eb17d56f5c8c54dd5e53372703744491c27

![screen shot 2017-06-01 at 09 59 48](https://cloud.githubusercontent.com/assets/1285805/26683211/138a8416-46b1-11e7-9ca9-09736fba2970.png)

**Comment by christophergandrud** [comment]
Relevant discussion at: https://github.com/IQSS/Zelig/issues/273

**Comment by christophergandrud** [comment]
I think `to_zelig` experimental version could be included in version 5.1-2.

It is fairly stable for the limited range of models it supports and doesn't seem to break anything. There is also documentation introduced in: 21171f7ef036db294142f18ebc7924850c144bdc

**Comment by christophergandrud** [comment]
Initial to_zelig merged into master and ready for CRAN release.
